docs/Developers/NextGen_Core.rst

```diff
@@ -0,0 +1,173 @@
+NextGen-Core
+============
+
+This document provides a brief overview of coala's NextGen-Core.
+coala's NextGen-Core comes with the promise of eliminating the dependencies
+amongst bears (by removing the queuing mechanism) and better performance
+(by utilizing asynchronous I/O and process pool).
+
+What has changed?
+-----------------
+
+``Global Bears`` are now called ``Project Bears`` and ``Local Bears`` are
+now known as ``File Bears``. Both the ``ProjectBear`` and the ``FileBear``
+classes inherit from the base class ``coalib.core.Bear.py``.
+
+- **ProjectBears:** Performs analysis on the whole codebase and does not
+  parallelize tasks. All the bears formerly known as GlobalBears will now
+  inherit from this class.
+- **FileBears:** Performs analysis on the files provided by the user
+  and parallelizes tasks for each give file. All the bears formerly known as
+  LocalBears will now inherit from this class.
+
+There is no need to pass ``HiddenResults`` between different bears (made
+possible by the new dependency resolution) to hide results to the result
+callback. Only the results that were explicitly requested by passing the needed
+bears are passed now.
+
+What is new?
+------------
+
+The following new features have been added as a part of the NextGen-Core:
+
+- New bear type: ``DependencyBear``
+- Usage of instance-bound ``Bear.BEAR_DEPS``
+- Improved dependency system
+- Custom parallelization for bears
+- Superior caching
+
+DependencyBear
+--------------
+
+The handling of bear dependencies by the old core wasn't effective. The
+old core used a queuing mechanism to communicate between bear runs which
+led to problems of a GlobalBear not running if any one of its
+dependencies (a local bear) terminated. The NextGen-Core does away with
+that by introducing a ``DependencyBear``.
```

Comment:
Actually `DependencyBear` is not lifting the limitation itself, the core does it ;) This bear is just a convenience for writing dependency bears, and has the advantage that it takes care of caching as well (by using the task objects to pass dependency results)